File: lib/domain/cultivation_plan/entities/cultivation_plan_entity.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Entities
      # 永続化された栽培計画（Adapter→Interactor 向けスナップショット）
      class CultivationPlanEntity
        attr_reader :id, :farm_id, :user_id, :total_area, :plan_type, :plan_year, :plan_name,
                    :planning_start_date, :planning_end_date, :status, :session_id, :display_name,
                    :cultivation_plan_crops_count, :cultivation_plan_fields_count, :updated_at, :created_at

        def initialize(
          id:,
          farm_id:,
          user_id:,
          total_area:,
          plan_type:,
          plan_year: nil,
          plan_name: nil,
          planning_start_date: nil,
          planning_end_date: nil,
          status: nil,
          session_id: nil,
          display_name: nil,
          cultivation_plan_crops_count: 0,
          cultivation_plan_fields_count: 0,
          created_at: nil,
          updated_at: nil
        )
          @id = id
          @farm_id = farm_id
          @user_id = user_id
          @total_area = total_area
          @plan_type = plan_type
          @plan_year = plan_year
          @plan_name = plan_name
          @planning_start_date = planning_start_date
          @planning_end_date = planning_end_date
          @status = status
          @session_id = session_id
          @display_name = display_name
          @cultivation_plan_crops_count = cultivation_plan_crops_count
          @cultivation_plan_fields_count = cultivation_plan_fields_count
          @created_at = created_at
          @updated_at = updated_at
        end

        def plan_type_private?
          plan_type == "private"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/entities/cultivation_plan_entity.rs
//! Ruby: `Domain::CultivationPlan::Entities::CultivationPlanEntity`

#[derive(Debug, Clone, PartialEq)]
pub struct CultivationPlanEntity {
    pub id: i64,
    pub farm_id: i64,
    pub user_id: i64,
    pub total_area: f64,
    pub plan_type: String,
    pub plan_year: Option<i32>,
    pub plan_name: Option<String>,
    pub planning_start_date: Option<String>,
    pub planning_end_date: Option<String>,
    pub status: Option<String>,
    pub session_id: Option<String>,
    pub display_name: Option<String>,
    pub optimization_phase: Option<String>,
    pub optimization_phase_message: Option<String>,
    pub cultivation_plan_crops_count: i32,
    pub cultivation_plan_fields_count: i32,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

impl CultivationPlanEntity {
    pub fn plan_type_private(&self) -> bool {
        self.plan_type == "private"
    }
}
